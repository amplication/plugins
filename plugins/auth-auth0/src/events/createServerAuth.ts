import {
  CreateServerAuthParams,
  DsgContext,
  ModuleMap,
} from "@amplication/code-gen-types";
import { serverStaticPath } from "../constants";
import {
  createAuthModule,
  createAuthResolver,
  createJwtStrategy,
  createJwtStrategyBase,
} from "../core";

export const beforeCreateAuthModules = (
  context: DsgContext,
  eventParams: CreateServerAuthParams
) => {
  context.utils.skipDefaultBehavior = true;
  return eventParams;
};

export const afterCreateAuthModules = async (
  context: DsgContext,
  eventParams: CreateServerAuthParams,
  modules: ModuleMap
) => {
  const staticFiles = await context.utils.importStaticModules(
    serverStaticPath,
    context.serverDirectories.srcDirectory
  );

  // 1. Create JWT strategy
  const { module: jwtStrategy, searchableAuthField } = await createJwtStrategy(
    context
  );
  modules.set(jwtStrategy);

  // 2. Create JWT strategy base
  const jwtStrategyBase = await createJwtStrategyBase(
    context,
    searchableAuthField
  );
  modules.set(jwtStrategyBase);

  // 3. Create auth module
  const authModule = await createAuthModule(context);
  modules.set(authModule);

  // 4. Create auth resolver
  const authResolver = await createAuthResolver(context);
  modules.set(authResolver);

  await modules.merge(staticFiles);

  // Remove array of files
  const filesToRemove: string[] = [
    "auth.controller.ts",
    "auth.service.ts",
    "auth.service.spec.ts",
    "constants.ts",
    "ITokenService.ts",
    "LoginArgs.ts",
    "password.service.ts",
    "password.service.spec.ts",
    "token.service.ts",
  ];
  modules.removeMany(
    filesToRemove.map(
      (file) => `${context.serverDirectories.authDirectory}/${file}`
    )
  );
  return modules;
};
