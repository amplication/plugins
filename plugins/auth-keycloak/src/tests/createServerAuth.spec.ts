import { mock } from "jest-mock-extended";
import fg from "fast-glob";
import * as fs from "fs";
import {
  DsgContext,
  ModuleMap,
  BuildLogger,
} from "@amplication/code-gen-types";
import { afterCreateAuthModules } from "@events/createServerAuth";

describe("Testing afterCreateApp hook", () => {
  let context: DsgContext;
  let modules: ModuleMap;
  let logger: BuildLogger;

  beforeEach(() => {
    logger = mock<BuildLogger>();
    context = mock<DsgContext>({
      resourceInfo: {
        settings: {
          authEntityName: "User",
        },
      },
      entities: [
        {
          name: "User",
        },
      ],
      logger: {
        info: jest.fn(),
        error: jest.fn(),
        warn: jest.fn(),
      },
      pluginInstallations: [
        {
          npm: "@amplication/plugin-auth-auth0",
        },
      ],
      utils: {
        importStaticModules: async (source, basePath) => {
          const directory = `${source}/`;
          const staticModules = await fg(`${directory}**/*`, {
            absolute: false,
            dot: true,
            ignore: ["**.js", "**.js.map", "**.d.ts"],
          });

          const modules = await Promise.all(
            staticModules.sort().map(async (module) => {
              return {
                path: module.replace(directory, basePath ? basePath + "/" : ""),
                code: await fs.promises.readFile(module, "utf-8"),
              };
            }),
          );

          const moduleMap: ModuleMap = new ModuleMap(logger);

          for await (const module of modules) {
            await moduleMap.set(module);
          }

          return moduleMap;
        },
      },
    });
    modules = new ModuleMap(logger);
  });

  it("should create all the auth modules", async () => {
    const newModules = await afterCreateAuthModules(context, {}, modules);

    expect(newModules.modules()).toMatchSnapshot();
  });

  it("should remove the default auth modules", async () => {
    const newModules = await afterCreateAuthModules(context, {}, modules);
    const filesToRemove: string[] = [
      "auth.controller.ts",
      "auth.service.ts",
      "auth.service.spec.ts",
      "constants.ts",
      "ITokenService.ts",
      "LoginArgs.ts",
      "password.service.ts",
      "password.service.spec.ts",
      "token.service.ts",
    ];

    filesToRemove.forEach((file) => {
      expect(newModules.get(file)).toBeFalsy();
    });
  });
});
