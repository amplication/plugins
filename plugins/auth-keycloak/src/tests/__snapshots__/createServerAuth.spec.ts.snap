// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing afterCreateApp hook should create all the auth modules 1`] = `
[
  {
    "code": "import { Injectable } from "@nestjs/common";
import { JwtStrategyBase } from "./base/jwt.strategy.base";
import { ConfigService } from "@nestjs/config";
import { Auth0User } from "./base/User";
import { IAuthStrategy } from "../IAuthStrategy";
import { UserInfo } from "../UserInfo";
import { UserService } from "src/user/user.service";
@Injectable()
export class JwtStrategy extends JwtStrategyBase implements IAuthStrategy {
  constructor(
    protected readonly configService: ConfigService,
    protected readonly userService: UserService
  ) {
    super(configService, userService);
  }
  async validate(payload: { user: Auth0User }): Promise<UserInfo> {
    const validatedUser = await this.validateBase(payload);
    // If the entity is valid, return it
    if (validatedUser) {
      return validatedUser;
    }
    // Otherwise, make a new entity and return it
    const userFields = payload.user;
    const { name, email } = userFields;
    const newUser = await this.userService.create({
      data: {
        name,
        email,
        ...userFields,
        // TODO: Add the rest of the fields here according to your authEntity schema
      },
    });
    return { ...newUser, roles: newUser?.roles as string[] };
  }
}
",
    "path": "undefined/jwt/jwt.strategy.ts",
  },
  {
    "code": "import { ConfigService } from "@nestjs/config";
import { PassportStrategy } from "@nestjs/passport";
import { passportJwtSecret } from "jwks-rsa";
import { ExtractJwt, Strategy } from "passport-jwt";
import { Auth0User } from "./User";
import { UserInfo } from "../../UserInfo";
import { UserService } from "src/user/user.service";
export class JwtStrategyBase extends PassportStrategy(Strategy) {
  constructor(
    protected readonly configService: ConfigService,
    protected readonly userService: UserService
  ) {
    super({
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(), // Extract JWT from the Authorization header
      audience: configService.get("AUTH0_AUDIENCE"), // The resource server where the JWT is processed
      issuer: \`\${configService.get("AUTH0_ISSUER_URL")}\`, // The issuing Auth0 server
      algorithms: ["RS256"], // Asymmetric signing algorithm
      secretOrKeyProvider: passportJwtSecret({
        cache: true,
        rateLimit: true,
        jwksRequestsPerMinute: 5,
        jwksUri: \`\${configService.get(
          "AUTH0_ISSUER_URL"
        )}.well-known/jwks.json\`,
      }),
    });
  }
  // Validate the received JWT and construct the user object out of the decoded token.
  async validateBase(payload: {
    user: Auth0User;
  }): Promise<UserInfo | null> {
    const { email, name } = payload.user;
    const user = await this.userService.findOne({
      where: {
        name_email: { email, name },
      },
    });
    return user ? { ...user, roles: user?.roles as string[] } : null;
  }
}
",
    "path": "undefined/jwt/base/jwt.strategy.base.ts",
  },
  {
    "code": "import { Module, forwardRef } from "@nestjs/common";
import { PassportModule } from "@nestjs/passport";
import { JwtStrategy } from "./jwt/jwt.strategy";
import { UserModule } from "../user/user.module";
@Module({
  imports: [
    forwardRef(() => UserModule),
    PassportModule.register({ defaultStrategy: "jwt" }),
  ],
  providers: [JwtStrategy],
  controllers: [],
  exports: [PassportModule],
})
export class AuthModule {}
",
    "path": "undefined/auth.module.ts",
  },
  {
    "code": "import * as common from "@nestjs/common";
import { Query, Resolver } from "@nestjs/graphql";
import * as gqlACGuard from "../auth/gqlAC.guard";
import { GqlDefaultAuthGuard } from "./gqlDefaultAuth.guard";
import { UserData } from "./userData.decorator";
import { UserInfo } from "./UserInfo";
@Resolver(UserInfo)
export class AuthResolver {
  @Query(() => UserInfo)
  @common.UseGuards(GqlDefaultAuthGuard, gqlACGuard.GqlACGuard)
  async userInfo(
    @UserData() entityInfo: UserInfo
  ): Promise<UserInfo> {
    return entityInfo;
  }
}
",
    "path": "undefined/auth.resolver.ts",
  },
  {
    "code": "import { Observable } from "rxjs";
import { ExecutionContext, Injectable } from "@nestjs/common";
import { Reflector } from "@nestjs/core";
import { IS_PUBLIC_KEY } from "../decorators/public.decorator";
import { JwtAuthGuard } from "./jwt/jwtAuth.guard";
@Injectable()
export class DefaultAuthGuard extends JwtAuthGuard {
  constructor(private readonly reflector: Reflector) {
    super();
  }
  canActivate(
    context: ExecutionContext
  ): boolean | Promise<boolean> | Observable<any> {
    const isPublic = this.reflector.get<boolean>(
      IS_PUBLIC_KEY,
      context.getHandler()
    );
    if (isPublic) {
      return true;
    }
    return super.canActivate(context);
  }
}
",
    "path": "auth/defaultAuth.guard.ts",
  },
  {
    "code": "export interface Auth0User {
  nickname: string;
  username: string;
  name: string;
  email: string;
  picture: string;
  given_name: string;
  family_name: string;
}
",
    "path": "auth/jwt/base/User.ts",
  },
  {
    "code": "import { ExecutionContext } from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { Observable } from "rxjs";
export class JwtAuthGuard extends AuthGuard("jwt") {
  canActivate(
    context: ExecutionContext
  ): boolean | Promise<boolean> | Observable<boolean> {
    const promiseOrBoolean = super.canActivate(context);
    if (promiseOrBoolean instanceof Promise) {
      return promiseOrBoolean.then((result) => {
        return result;
      });
    } else {
      return promiseOrBoolean;
    }
  }
}
",
    "path": "auth/jwt/jwtAuth.guard.ts",
  },
]
`;